### PR TITLE
Check for forced attribute changes

### LINKED ISSUE #27956: Regression - attribute_will_change! doesn't work for nested resources
### Failing Test

```ruby
begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"
  gem "rails", github: "rails/rails"
  gem "arel", github: "rails/arel"
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"

ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")

ActiveRecord::Schema.define do
  create_table :users, force: true do |t|
    t.string :name
  end

  create_table :posts, force: true do |t|
    t.integer :user_id
    t.text :body
  end
end

class User < ActiveRecord::Base
  has_many :posts
  accepts_nested_attributes_for :posts
end

class Post < ActiveRecord::Base
  attr_accessor :virtual_input_attribute

  belongs_to :user
  before_save :set_post_body

  def virtual_input_attribute=(value)
    attribute_will_change!(:virtual_input_attribute) unless value.blank?
    @virtual_input_attribute = value
  end

  def set_post_body
    self.body = virtual_input_attribute
  end
end

p ActiveRecord::VERSION::STRING

class BugTest < Minitest::Test
  # Fails on edge Rails. Passes on 5.0.1
  def test_nested_resource_dirty
    user = User.create! name: 'Ruby'
    post = Post.create!(user_id: user.id, body: nil)

    user.update(posts_attributes: { '0' => { id: post.id, virtual_input_attribute: '1' } })

    assert_equal '1', post.reload.body
  end
end
```

### Expected behavior
attribute_will_change! should make nested resources dirty, and the post's body should be updated.

### Actual behavior
attribute_will_change! has no effect for nested resources. The post's body remains nil.

It still works when directly updating the resource. This test passes:

```ruby
  def test_resource_dirty
    user = User.create! name: 'Ruby'
    post = Post.create!(user_id: user.id, body: nil)
    
    post.update(virtual_input_attribute: '1')

    assert_equal '1', post.reload.body
  end
```

### System configuration
**Rails version**:
Edge Rails cabb212976703dac46bacc36addf70a232e96e71
**Ruby version**:
ruby 2.4.0p0

I tracked it down to this commit: https://github.com/rails/rails/commit/16ae3db5a5c6a08383b974ae6c96faac5b4a3c81

### PR DESCRIPTION
**Pull request by alse**
### Summary

AttributeMutationTracker#any_changes? only looked at model attribute
changes, and ignored any forced changes. Fixes #27956

### CODE REVIEW

File: activerecord/test/models/parrot.rb [hunk 1]
```diff
@@ -18,6 +18,18 @@ def cancel_save_callback_method
...
+  remove_method "new_name_input="
+  def new_name_input=(value)
+    attribute_will_change!(:new_name_input)
+    @new_name_input = value
+  end
+
+  before_update :set_new_name_from_input, if: :new_name_input
+  def set_new_name_from_input
+    self.name = new_name_input
+  end
```
Comment (sgrif) [inline activerecord/test/models/parrot.rb]: Is there a simpler way we can test this? Perhaps just `parrot.name_will_change!; assert parrot.changed?`?

**Review by sgrif — CHANGES_REQUESTED** [review]
I'm fine with the code change, but the test feels very off to me.

File: activerecord/test/models/parrot.rb (hunk 1 above)
Comment (alse) [inline activerecord/test/models/parrot.rb]: You're right, I tried to simplify the test. The changed? method already accounts for changes to virtual attributes, so I used has_changes_to_save? instead.

File: activerecord/lib/active_record/attribute_mutation_tracker.rb [hunk 2]
```diff
@@ -31,7 +31,7 @@ def change_to_attribute(attr_name)
     end
 
     def any_changes?
-      attr_names.any? { |attr| changed?(attr) }
+      attr_names.any? { |attr| changed?(attr) } || forced_changes.any?
```
Comment (alse) [inline activerecord/lib/active_record/attribute_mutation_tracker.rb]: Another way to fix this issue is to include virtual attributes in the attr_names. The changed? method already checks for forced changes

Comment (sgrif) [inline activerecord/lib/active_record/attribute_mutation_tracker.rb]: By "virtual attribute" are you referring to an `attr_accessor`? Those are not intended to work with dirty.

**Comment by sgrif** [comment]
I looked more closely at this. This is not a bug.

**Comment by sgrif** [comment]
Hm. I saw that this worked in Rails 5.0.1. I thought I had changed it then.

**Comment by shioyama** [comment]
It might be pointless to contest this change, but I'm going to try anyway.

I have a gem ([Mobility](https://github.com/shioyama/mobility)) which adds (virtual) translated attributes to a model, and one of the things I do is to support [dirty tracking](https://github.com/shioyama/mobility/blob/master/lib/mobility/backend/active_model/dirty.rb) on these virtual attributes (which may be stored in different formats in the db depending on the choice of backend).

Up until now, I have been able to mark changes using e.g. `title_en` for a change on the title in English, even though `title_en` is not a real attribute, and these changes would appear with changes to other ("real") attributes. This use of `attr_will_change!` is completely consistent with Active*Model* Dirty, and I found it very simple to apply to my situation.

Of course Active*Record* Dirty overrides all these methods and does slightly different things, but I managed to get around this with a minor [patch](https://github.com/lime/mobility/blob/64df13697927d884c17e571e539fc8740235266c/lib/mobility/active_record.rb#L15-L27).

But now, with this change, you are requiring me to declare all these attributes *upfront*. This essentially means that you can no longer leverage the flexibility of Active*Model* Dirty within an ActiveRecord model, because ActiveRecord overrides all these dirty methods. It almost seems like you might as well not even include `ActiveModel::Dirty` in `ActiveRecord::Dirty` anymore since the two at this point are so dissimilar.

I understand that #27956 was probably the kind of issue you don't want to deal with, but deprecating this "open-ended" use of attributes in AR models appears to be something that was already [in the plans](https://github.com/rails/rails/issues/27956#issuecomment-279179849). All things being equal, I would think that imposing as few requirements as possible would be the ideal to strive for. In my particular case, the change will probably require an actual change in how my gem works, because I do not want to have to specify all locales upfront (which I would need in order to specify the attributes).

Any chance this could be reconsidered, or alternatively could you explain why this is necessary in Dirty in particular? My preference woudl be for AR::Dirty to as much as possible stick to AM::Dirty, for consistency and flexibility's sake.

**Pull request closed without merging**